### PR TITLE
[NeMo UX] Support generating datasets using different train/valid/test distributions

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/gpt/data/pre_training.py [hunk 1]
```diff
@@ -92,8 +106,12 @@ def setup(self, stage: str = "") -> None:
         num_test_samples = int(test_iters * self.data_sampler.global_batch_size)
 
         if self.trainer.limit_val_batches <= 1.0 and isinstance(self.trainer.limit_val_batches, float):
+            assert (
+                "blend" not in self.build_kwargs
+            ), f"When using a single data distribution, limit_val_batches <= 1.0 is not supported."
+
```
Comment (athitten) [inline nemo/collections/llm/gpt/data/pre_training.py:112]: Also I think this comment can be confusing to users. Can we elaborate what it means by single data distribution. Basically clarify that there is no split of "train", "valid", "test" in the dataset passed and float limit_val_batches isn't supported in that case.

File: nemo/collections/llm/gpt/data/pre_training.py [hunk 2]
```diff
@@ -38,16 +38,30 @@ def __init__(
...
+        build_kwargs = {}
+        if isinstance(paths, dict):
+            if split is not None:
+                warnings.warn(
+                    f"{split=} will be ignored since datasets are being created " f"from 3 separate distributions."
+                )
+            build_kwargs["blend_per_split"] = [
+                get_blend_from_list(paths["train"]),
+                get_blend_from_list(paths["validation"]),
+                get_blend_from_list(paths["test"]),
```
Comment (athitten) [inline nemo/collections/llm/gpt/data/pre_training.py:55]: @ashors1 does get_blend_from_list(paths["train"]) work even when you have multiple data files. For ex: {"train": /datafile1/, /datafile2/}. Also in this case if weights are ignored then is the dataset built with all samples from both /datafile1/ and /datafile2/ ?

Comment (ashors1) [reply nemo/collections/llm/gpt/data/pre_training.py:55]: Yes, `get_blend_from_list(paths["train"])` works when you have multiple paths. You're also able to pass in weights by interleaving them with the paths. For example, the following would work: 
```
paths={
            "train": [25, PATH1, 75, PATH2],
            "validation": [PATH3, PATH4],
            "test": ['1', PATH5], 
        }
```

The only time the weights are not used is when `limit_val_batches <= 1.0`, in which case we want to return the full validation dataset. In this case, users are expected not to provide weights for the paths.

**Review by athitten — APPROVED** [review]
LGTM, thanks @ashors1! Also for future references might be good to document the behavior of the megatron datasets as to what is supported and what is not.

**Pull request merged**